src/PlaneWaveBasis.jl

```diff
@@ -245,13 +246,58 @@ build_kpoints(basis::PlaneWaveBasis, kcoords) =
     basis
 end
 
+# This is the "internal" constructor; the higher-level one below should be preferred
+@timing function PlaneWaveBasis(model::Model{T}, Ecut::Number,
+                                kcoords::AbstractVector, ksymops, symmetries=nothing;
+                                fft_size=nothing, variational=true,
+                                optimize_fft_size=false, supersampling=2,
+                                kgrid=nothing, kshift=nothing,
+                                comm_kpts=MPI.COMM_WORLD,
+                               ) where {T <: Real}
+    mpi_ensure_initialized()
+
+    # Compute or validate fft_size
+    if fft_size === nothing
+        fft_size = compute_fft_size(model::Model{T}, Ecut, supersampling,
+                                    variational, optimize_fft_size, kcoords)
+    else
+        # validate
+        if variational
+            max_E = sum(abs2, model.recip_lattice * floor.(Int, Vec3(fft_size) ./ 2)) / 2
+            Ecut > max_E && @warn(
+                "For a variational method, Ecut should be less than the maximal kinetic " *
+                    "energy the grid supports ($max_E)"
+            )
+        else
+            # ensure no other options are set
+            @assert supersampling == 2
+            @assert !optimize_fft_size
+        end
+    end
+
+    # Validate kpoints and symmetries
+    @assert length(kcoords) == length(ksymops)
+    if symmetries === nothing
+        # TODO instead compute the group generated by ksymops (also in
+        # the other constructors). Not critical because this should
+        # not be used in this context anyway...
+        symmetries = vcat(ksymops...)
+    end
+    PlaneWaveBasis(model, Ecut,
+                   kcoords, ksymops,
+                   fft_size, variational, symmetries,
+                   kgrid, kshift, comm_kpts)
+end
+
 """
-Creates a new basis identical to `basis`, but with a different set of kpoints
+Creates a new basis identical to `basis`, but with a custom set of kpoints
 """
 function PlaneWaveBasis(basis::PlaneWaveBasis, kcoords::AbstractVector,
-                        ksymops::AbstractVector, symmetries=nothing)
-    PlaneWaveBasis(basis.model, basis.Ecut, kcoords, ksymops, symmetries;
-                   fft_size=basis.fft_size, variational=basis.variational)
+                        ksymops::AbstractVector, symmetries=vcat(ksymops...))
+    PlaneWaveBasis(basis.model, basis.Ecut, kcoords, ksymops,
+                   basis.fft_size, basis.variational, basis.symmetries,
+                   nothing, nothing, # explicitly given kpoints
```

Comment:
I think for clarity I would assign these to variables `kgrid` and `kshift` and pass them. That's one line more and we don't forget what `nothing` refers to. Also easier for later refactoring.

Reply:
good point!

Reply:
well that's one of the reasons why I dislike "true", "false" or "nothing" as positional arguments ...